pypulseq/make_extended_trapezoid_area.py

```diff
@@ -35,100 +33,93 @@ def make_extended_trapezoid_area(
+    # try different starting points for the optimization
+    optimization_results = [
+        minimize(fun=obj1, x0=-max_grad, method="Nelder-Mead"),
+        minimize(fun=obj1, x0=0, method="Nelder-Mead"),
+        minimize(fun=obj1, x0=max_grad, method="Nelder-Mead"),
+    ]
+    # choose the best result
+    optimization_results = np.array([(*res.x, res.fun) for res in optimization_results])
+    grad_amp_results, obj1val_results = (
+        optimization_results[:, 0],
+        optimization_results[:, 1],
+    )
+    best_index = np.argmin(obj1val_results)
+    grad_amp = grad_amp_results[best_index]
+    obj1_value = obj1val_results[best_index]
+
+    if obj1_value**0.5 > 1e-3 or abs(grad_amp) > max_grad:
+        # Optimization without flat_time did not converge. Try to find flat_time.
+        grad_amp = math.copysign(max_grad, grad_amp)
+        obj2 = lambda x: (area - _calc_area_and_times(grad_amp, x, slew_rate, grad_start, grad_end)) ** 2
+        res2 = minimize(fun=obj2, x0=system.grad_raster_time, method="Nelder-Mead")
+        flat_time = res2.x[0]
+        # assert flat_time is larger than 0 and that the objective function is small
+        assert flat_time >= 0, "flat time is smaller than 0, this should not happen"
+
+    # ensure times are on raster
+    flat_time = _to_raster(flat_time)
+    time_ramp_up = _to_raster(_calc_ramp_up_time(grad_amp, slew_rate, grad_start))
+    time_ramp_down = _to_raster(_calc_ramp_down_time(grad_amp, slew_rate, grad_end))
+
+    # recalculate grad_amp while taking raster effect into account
+    obj3 = lambda x: (area - _calc_area(x, time_ramp_up, flat_time, time_ramp_down, grad_start, grad_end)) ** 2
+    res = minimize(fun=obj3, x0=grad_amp, method="Nelder-Mead")
+    grad_amp = res.x[0]
+
+    if flat_time > 0:
+        times = cumsum(0, time_ramp_up, flat_time, time_ramp_down)
+        amplitudes = [grad_start, grad_amp, grad_amp, grad_end]
     else:
-        Tru = (
-            math.ceil(abs(Gp - grad_start) / SR / system.grad_raster_time)
-            * system.grad_raster_time
-        )
-        Trd = (
-            math.ceil(abs(Gp - grad_end) / SR / system.grad_raster_time)
-            * system.grad_raster_time
-        )
-
-        if Trd > 0:
-            if Tru > 0:
-                times = cumsum(0, Tru, Trd)
-                amplitudes = np.array([grad_start, Gp, grad_end])
+        if time_ramp_down > 0:
+            if time_ramp_up > 0:
+                times = cumsum(0, time_ramp_up, time_ramp_down)
+                amplitudes = np.array([grad_start, grad_amp, grad_end])
             else:
-                times = cumsum(0, Trd)
+                times = cumsum(0, time_ramp_down)
                 amplitudes = np.array([grad_start, grad_end])
         else:
-            times = cumsum(0, Tru)
+            times = cumsum(0, time_ramp_up)
             amplitudes = np.array([grad_start, grad_end])
```

Comment:
- The `flat_time > 0` section needs checks for `time_ramp_up` and/or `time_ramp_down` being equal to 0.
- The checks may need to consider rounding errors?

Related error:
```pp.make_extended_trapezoid_area(1000, 'x', system.max_grad*0.99, 0, system)```